## Part 1

		1. Source three years of price data from a random set of five equities

		2. Store in a dataframe

		3. Write a procedure to iterate through the dataframe to simulate daily price stream and print the prices

		4. Writing a class to store a single unit of price data (timestamp, open, high, low, close, volume)

		5. Run the iteration procedure and for each security, save the data into a dictionary indexed by datetime

## Part 2

		1. Install MongoDB Atlas on your laptop (https://docs.atlas.mongodb.com/)
        
		2. Download the sample data onto your free cluster

		3. Set up a notebook to extract some of the data

In [1]:
import pandas as pd
import pandas_datareader.data as web
import datetime 
import numpy as np
import matplotlib.pyplot as plt

In [2]:
%matplotlib inline

In [3]:
pd.set_option("max_rows", 10)
pd.set_option("min_rows", 8)

In [4]:
pd.set_option("display.notebook_repr_html", False)
pd.set_option("display.max_columns", 12)
pd.set_option("display.width", 80)
pd.set_option("precision", 3)

## Part 1
### Q1 & Q2  Source three years of price data from a random set of five equities & Store in a dataframe
> We decided to focus on the 5 FAANG stocks

- Facebook, Amazon, Apple, Netflix and Alphabet (formerly known as Google).

- Data of these 5 stocks were stored in a DataFrame.

- Start date of our data: 30 Sep 2017
- End date of our data: 1 Oct 2020

In [5]:
start = datetime.date(2017, 9, 30)
end = datetime.date(2020, 10, 1)

In [6]:
def get(tickers, start, end):
    
    def data(ticker):
        return web.DataReader(ticker, "yahoo", start, end)
    
    stocks = map(data, tickers)
    
    return pd.concat(stocks, keys = tickers, names = ["Ticker", "Date"])

In [7]:
tickers = ["FB", "AMZN", "AAPL", "NFLX", "GOOGL"]

In [8]:
faang = get(tickers, start, end)

In [9]:
faang

                      High      Low     Open    Close     Volume  Adj Close
Ticker Date                                                                
FB     2017-09-29   171.66   168.81   168.83   170.87  1.534e+07     170.87
       2017-10-02   171.87   168.75   171.39   169.47  1.339e+07     169.47
       2017-10-03   170.73   169.20   169.30   169.96  8.190e+06     169.96
       2017-10-04   170.67   168.29   169.83   168.42  1.213e+07     168.42
...                    ...      ...      ...      ...        ...        ...
GOOGL  2020-09-28  1472.99  1446.28  1471.99  1458.66  1.539e+06    1458.66
       2020-09-29  1472.00  1454.00  1463.91  1466.02  1.591e+06    1466.02
       2020-09-30  1486.20  1455.76  1460.94  1465.60  2.045e+06    1465.60
       2020-10-01  1495.07  1476.80  1483.66  1487.90  1.644e+06    1487.90

[3785 rows x 6 columns]

Our data consists of 3785 rows and 6 columns.

The columns include, for each of the FAANG stock, the following:

- Adjusted Close
- Close
- High
- Low
- Open
- Volume

In [10]:
type(faang)

pandas.core.frame.DataFrame

In [11]:
print("Shape of Dataframe: {}".format(faang.shape))
print("Column Names: {}".format(faang.columns))

Shape of Dataframe: (3785, 6)
Column Names: Index(['High', 'Low', 'Open', 'Close', 'Volume', 'Adj Close'], dtype='object')


### Q3 Procedure to Iterate through the Dataframe to Simulate Daily Price Stream and Print the Prices

In [12]:
AAPL = faang.loc['AAPL',:]
AAPL_df = AAPL.reset_index()

GOOGL = faang.loc['GOOGL',:]
GOOGL_df = GOOGL.reset_index()

NFLX = faang.loc['NFLX',:]
NFLX_df = NFLX.reset_index()

FB = faang.loc['FB',:]
FB_df = FB.reset_index()

AMZN = faang.loc['AMZN',:]
AMZN_df = AMZN.reset_index()

In [13]:
# print out the price of Apple Stock in AAPL dataframe
for row in AAPL.iterrows():
    print("On {} the adjusted closing price was ${:.2f}.".format(row[0].date(), row[1]['Adj Close']))

On 2017-09-29 the adjusted closing price was $36.65.
On 2017-10-02 the adjusted closing price was $36.57.
On 2017-10-03 the adjusted closing price was $36.73.
On 2017-10-04 the adjusted closing price was $36.49.
On 2017-10-05 the adjusted closing price was $36.95.
On 2017-10-06 the adjusted closing price was $36.93.
On 2017-10-09 the adjusted closing price was $37.05.
On 2017-10-10 the adjusted closing price was $37.07.
On 2017-10-11 the adjusted closing price was $37.22.
On 2017-10-12 the adjusted closing price was $37.09.
On 2017-10-13 the adjusted closing price was $37.33.
On 2017-10-16 the adjusted closing price was $38.01.
On 2017-10-17 the adjusted closing price was $38.15.
On 2017-10-18 the adjusted closing price was $37.99.
On 2017-10-19 the adjusted closing price was $37.09.
On 2017-10-20 the adjusted closing price was $37.15.
On 2017-10-23 the adjusted closing price was $37.13.
On 2017-10-24 the adjusted closing price was $37.35.
On 2017-10-25 the adjusted closing price was $

In [14]:
# print out the price of Facebook Stock in FB dataframe
for row in FB.iterrows():
    print("On {} the adjusted closing price was ${:.2f}.".format(row[0].date(), row[1]['Adj Close']))

On 2017-09-29 the adjusted closing price was $170.87.
On 2017-10-02 the adjusted closing price was $169.47.
On 2017-10-03 the adjusted closing price was $169.96.
On 2017-10-04 the adjusted closing price was $168.42.
On 2017-10-05 the adjusted closing price was $171.24.
On 2017-10-06 the adjusted closing price was $172.23.
On 2017-10-09 the adjusted closing price was $172.50.
On 2017-10-10 the adjusted closing price was $171.59.
On 2017-10-11 the adjusted closing price was $172.74.
On 2017-10-12 the adjusted closing price was $172.55.
On 2017-10-13 the adjusted closing price was $173.74.
On 2017-10-16 the adjusted closing price was $174.52.
On 2017-10-17 the adjusted closing price was $176.11.
On 2017-10-18 the adjusted closing price was $176.03.
On 2017-10-19 the adjusted closing price was $174.56.
On 2017-10-20 the adjusted closing price was $174.98.
On 2017-10-23 the adjusted closing price was $171.27.
On 2017-10-24 the adjusted closing price was $171.80.
On 2017-10-25 the adjusted c

On 2019-05-09 the adjusted closing price was $188.65.
On 2019-05-10 the adjusted closing price was $188.34.
On 2019-05-13 the adjusted closing price was $181.54.
On 2019-05-14 the adjusted closing price was $180.73.
On 2019-05-15 the adjusted closing price was $186.27.
On 2019-05-16 the adjusted closing price was $186.99.
On 2019-05-17 the adjusted closing price was $185.30.
On 2019-05-20 the adjusted closing price was $182.72.
On 2019-05-21 the adjusted closing price was $184.82.
On 2019-05-22 the adjusted closing price was $185.32.
On 2019-05-23 the adjusted closing price was $180.87.
On 2019-05-24 the adjusted closing price was $181.06.
On 2019-05-28 the adjusted closing price was $184.31.
On 2019-05-29 the adjusted closing price was $182.19.
On 2019-05-30 the adjusted closing price was $183.01.
On 2019-05-31 the adjusted closing price was $177.47.
On 2019-06-03 the adjusted closing price was $164.15.
On 2019-06-04 the adjusted closing price was $167.50.
On 2019-06-05 the adjusted c

In [15]:
# print out the price of Amazon Stock in AAPL dataframe
for row in AMZN.iterrows():
    print("On {} the adjusted closing price was ${:.2f}.".format(row[0].date(), row[1]['Adj Close']))

On 2017-09-29 the adjusted closing price was $961.35.
On 2017-10-02 the adjusted closing price was $959.19.
On 2017-10-03 the adjusted closing price was $957.10.
On 2017-10-04 the adjusted closing price was $965.45.
On 2017-10-05 the adjusted closing price was $980.85.
On 2017-10-06 the adjusted closing price was $989.58.
On 2017-10-09 the adjusted closing price was $990.99.
On 2017-10-10 the adjusted closing price was $987.20.
On 2017-10-11 the adjusted closing price was $995.00.
On 2017-10-12 the adjusted closing price was $1000.93.
On 2017-10-13 the adjusted closing price was $1002.94.
On 2017-10-16 the adjusted closing price was $1006.34.
On 2017-10-17 the adjusted closing price was $1009.13.
On 2017-10-18 the adjusted closing price was $997.00.
On 2017-10-19 the adjusted closing price was $986.61.
On 2017-10-20 the adjusted closing price was $982.91.
On 2017-10-23 the adjusted closing price was $966.30.
On 2017-10-24 the adjusted closing price was $975.90.
On 2017-10-25 the adjust

On 2019-02-13 the adjusted closing price was $1640.00.
On 2019-02-14 the adjusted closing price was $1622.65.
On 2019-02-15 the adjusted closing price was $1607.95.
On 2019-02-19 the adjusted closing price was $1627.58.
On 2019-02-20 the adjusted closing price was $1622.10.
On 2019-02-21 the adjusted closing price was $1619.44.
On 2019-02-22 the adjusted closing price was $1631.56.
On 2019-02-25 the adjusted closing price was $1633.00.
On 2019-02-26 the adjusted closing price was $1636.40.
On 2019-02-27 the adjusted closing price was $1641.09.
On 2019-02-28 the adjusted closing price was $1639.83.
On 2019-03-01 the adjusted closing price was $1671.73.
On 2019-03-04 the adjusted closing price was $1696.17.
On 2019-03-05 the adjusted closing price was $1692.43.
On 2019-03-06 the adjusted closing price was $1668.95.
On 2019-03-07 the adjusted closing price was $1625.95.
On 2019-03-08 the adjusted closing price was $1620.80.
On 2019-03-11 the adjusted closing price was $1670.62.
On 2019-03

On 2020-02-25 the adjusted closing price was $1972.74.
On 2020-02-26 the adjusted closing price was $1979.59.
On 2020-02-27 the adjusted closing price was $1884.30.
On 2020-02-28 the adjusted closing price was $1883.75.
On 2020-03-02 the adjusted closing price was $1953.95.
On 2020-03-03 the adjusted closing price was $1908.99.
On 2020-03-04 the adjusted closing price was $1975.83.
On 2020-03-05 the adjusted closing price was $1924.03.
On 2020-03-06 the adjusted closing price was $1901.09.
On 2020-03-09 the adjusted closing price was $1800.61.
On 2020-03-10 the adjusted closing price was $1891.82.
On 2020-03-11 the adjusted closing price was $1820.86.
On 2020-03-12 the adjusted closing price was $1676.61.
On 2020-03-13 the adjusted closing price was $1785.00.
On 2020-03-16 the adjusted closing price was $1689.15.
On 2020-03-17 the adjusted closing price was $1807.84.
On 2020-03-18 the adjusted closing price was $1830.00.
On 2020-03-19 the adjusted closing price was $1880.93.
On 2020-03

In [16]:
# print out the price of Netflix Stock in AAPL dataframe
for row in NFLX.iterrows():
    print("On {} the adjusted closing price was ${:.2f}.".format(row[0].date(), row[1]['Adj Close']))

On 2017-09-29 the adjusted closing price was $181.35.
On 2017-10-02 the adjusted closing price was $177.01.
On 2017-10-03 the adjusted closing price was $179.19.
On 2017-10-04 the adjusted closing price was $184.45.
On 2017-10-05 the adjusted closing price was $194.39.
On 2017-10-06 the adjusted closing price was $198.02.
On 2017-10-09 the adjusted closing price was $196.87.
On 2017-10-10 the adjusted closing price was $195.08.
On 2017-10-11 the adjusted closing price was $194.95.
On 2017-10-12 the adjusted closing price was $195.86.
On 2017-10-13 the adjusted closing price was $199.49.
On 2017-10-16 the adjusted closing price was $202.68.
On 2017-10-17 the adjusted closing price was $199.48.
On 2017-10-18 the adjusted closing price was $195.54.
On 2017-10-19 the adjusted closing price was $195.13.
On 2017-10-20 the adjusted closing price was $194.16.
On 2017-10-23 the adjusted closing price was $192.47.
On 2017-10-24 the adjusted closing price was $196.02.
On 2017-10-25 the adjusted c

On 2019-06-06 the adjusted closing price was $357.13.
On 2019-06-07 the adjusted closing price was $360.87.
On 2019-06-10 the adjusted closing price was $352.01.
On 2019-06-11 the adjusted closing price was $351.27.
On 2019-06-12 the adjusted closing price was $345.56.
On 2019-06-13 the adjusted closing price was $343.43.
On 2019-06-14 the adjusted closing price was $339.73.
On 2019-06-17 the adjusted closing price was $350.62.
On 2019-06-18 the adjusted closing price was $357.12.
On 2019-06-19 the adjusted closing price was $363.52.
On 2019-06-20 the adjusted closing price was $365.21.
On 2019-06-21 the adjusted closing price was $369.21.
On 2019-06-24 the adjusted closing price was $371.04.
On 2019-06-25 the adjusted closing price was $360.30.
On 2019-06-26 the adjusted closing price was $362.20.
On 2019-06-27 the adjusted closing price was $370.02.
On 2019-06-28 the adjusted closing price was $367.32.
On 2019-07-01 the adjusted closing price was $374.60.
On 2019-07-02 the adjusted c

On 2020-03-19 the adjusted closing price was $332.03.
On 2020-03-20 the adjusted closing price was $332.83.
On 2020-03-23 the adjusted closing price was $360.27.
On 2020-03-24 the adjusted closing price was $357.32.
On 2020-03-25 the adjusted closing price was $342.39.
On 2020-03-26 the adjusted closing price was $362.99.
On 2020-03-27 the adjusted closing price was $357.12.
On 2020-03-30 the adjusted closing price was $370.96.
On 2020-03-31 the adjusted closing price was $375.50.
On 2020-04-01 the adjusted closing price was $364.08.
On 2020-04-02 the adjusted closing price was $370.08.
On 2020-04-03 the adjusted closing price was $361.76.
On 2020-04-06 the adjusted closing price was $379.96.
On 2020-04-07 the adjusted closing price was $372.28.
On 2020-04-08 the adjusted closing price was $371.12.
On 2020-04-09 the adjusted closing price was $370.72.
On 2020-04-13 the adjusted closing price was $396.72.
On 2020-04-14 the adjusted closing price was $413.55.
On 2020-04-15 the adjusted c

In [17]:
# print out the price of Google Stock in AAPL dataframe
for row in GOOGL.iterrows():
    print("On {} the adjusted closing price was ${:.2f}.".format(row[0].date(), row[1]['Adj Close']))

On 2017-09-29 the adjusted closing price was $973.72.
On 2017-10-02 the adjusted closing price was $967.47.
On 2017-10-03 the adjusted closing price was $972.08.
On 2017-10-04 the adjusted closing price was $966.78.
On 2017-10-05 the adjusted closing price was $985.19.
On 2017-10-06 the adjusted closing price was $993.64.
On 2017-10-09 the adjusted closing price was $992.31.
On 2017-10-10 the adjusted closing price was $987.80.
On 2017-10-11 the adjusted closing price was $1005.65.
On 2017-10-12 the adjusted closing price was $1005.65.
On 2017-10-13 the adjusted closing price was $1007.87.
On 2017-10-16 the adjusted closing price was $1009.35.
On 2017-10-17 the adjusted closing price was $1011.00.
On 2017-10-18 the adjusted closing price was $1012.74.
On 2017-10-19 the adjusted closing price was $1001.84.
On 2017-10-20 the adjusted closing price was $1005.07.
On 2017-10-23 the adjusted closing price was $985.54.
On 2017-10-24 the adjusted closing price was $988.49.
On 2017-10-25 the ad

On 2018-11-27 the adjusted closing price was $1052.28.
On 2018-11-28 the adjusted closing price was $1091.79.
On 2018-11-29 the adjusted closing price was $1094.58.
On 2018-11-30 the adjusted closing price was $1109.65.
On 2018-12-03 the adjusted closing price was $1116.36.
On 2018-12-04 the adjusted closing price was $1062.47.
On 2018-12-06 the adjusted closing price was $1078.08.
On 2018-12-07 the adjusted closing price was $1046.58.
On 2018-12-10 the adjusted closing price was $1053.18.
On 2018-12-11 the adjusted closing price was $1061.65.
On 2018-12-12 the adjusted closing price was $1073.73.
On 2018-12-13 the adjusted closing price was $1073.54.
On 2018-12-14 the adjusted closing price was $1051.71.
On 2018-12-17 the adjusted closing price was $1025.65.
On 2018-12-18 the adjusted closing price was $1043.41.
On 2018-12-19 the adjusted closing price was $1035.46.
On 2018-12-20 the adjusted closing price was $1023.58.
On 2018-12-21 the adjusted closing price was $991.25.
On 2018-12-

On 2020-05-27 the adjusted closing price was $1420.28.
On 2020-05-28 the adjusted closing price was $1418.24.
On 2020-05-29 the adjusted closing price was $1433.52.
On 2020-06-01 the adjusted closing price was $1434.87.
On 2020-06-02 the adjusted closing price was $1442.31.
On 2020-06-03 the adjusted closing price was $1439.25.
On 2020-06-04 the adjusted closing price was $1414.30.
On 2020-06-05 the adjusted closing price was $1440.02.
On 2020-06-08 the adjusted closing price was $1448.04.
On 2020-06-09 the adjusted closing price was $1452.08.
On 2020-06-10 the adjusted closing price was $1464.70.
On 2020-06-11 the adjusted closing price was $1401.90.
On 2020-06-12 the adjusted closing price was $1412.92.
On 2020-06-15 the adjusted closing price was $1420.74.
On 2020-06-16 the adjusted closing price was $1446.47.
On 2020-06-17 the adjusted closing price was $1452.54.
On 2020-06-18 the adjusted closing price was $1434.12.
On 2020-06-19 the adjusted closing price was $1424.64.
On 2020-06

### Q4 Writing a class to store a single unit of price data (timestamp, open, high, low, close, volume)

In [18]:
faang.info

<bound method DataFrame.info of                       High      Low     Open    Close     Volume  Adj Close
Ticker Date                                                                
FB     2017-09-29   171.66   168.81   168.83   170.87  1.534e+07     170.87
       2017-10-02   171.87   168.75   171.39   169.47  1.339e+07     169.47
       2017-10-03   170.73   169.20   169.30   169.96  8.190e+06     169.96
       2017-10-04   170.67   168.29   169.83   168.42  1.213e+07     168.42
...                    ...      ...      ...      ...        ...        ...
GOOGL  2020-09-28  1472.99  1446.28  1471.99  1458.66  1.539e+06    1458.66
       2020-09-29  1472.00  1454.00  1463.91  1466.02  1.591e+06    1466.02
       2020-09-30  1486.20  1455.76  1460.94  1465.60  2.045e+06    1465.60
       2020-10-01  1495.07  1476.80  1483.66  1487.90  1.644e+06    1487.90

[3785 rows x 6 columns]>

In [19]:
faang.iloc[1]

High         1.719e+02
Low          1.688e+02
Open         1.714e+02
Close        1.695e+02
Volume       1.339e+07
Adj Close    1.695e+02
Name: (FB, 2017-10-02 00:00:00), dtype: float64

In [20]:
# Define a Class called "Stock" to store daily stock data
class stock_list:
    
    # Constructor
    def __init__(self, index,  openp, highp, lowp, closep, vol, adj_close):
        # Instance variables
        self.index = index
        self.openp = openp
        self.highp = highp
        self.lowp = lowp
        self.closep = closep
        self.vol = vol
        self.adj_close = adj_close
    
    # Instance Method
    def successful_load(self, run=False):
        if(run == False):
            pass
        else:
            print("Loaded Data {} Successfully.".format(self.index)) 

In [21]:
# List to store all the created stock objects
faang_list = []


for row in faang.iterrows():
    
    # Extracting all the features required to build the stock object
    index = row[0]
    openp = row[1].loc["Open"]
    highp = row[1].loc["High"]
    lowp = row[1].loc["Low"]
    closep = row[1].loc["Close"]
    vol = row[1].loc["Volume"]
    adj_close = row[1].loc["Adj Close"]
    
    # Create stock object
    faang_obj = stock_list(index, openp, highp, lowp, closep, vol, adj_close)
    faang_list.append(faang_obj)
    
    faang_obj.successful_load(run=True)

Loaded Data ('FB', Timestamp('2017-09-29 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-02 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-03 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-04 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-05 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-06 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-09 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-10 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-11 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-12 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-13 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-16 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-17 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-18 00:00:00')) Successfully.
Loaded Data ('FB', Timestamp('2017-10-19 00:00:00')) Successfu

Loaded Data ('AMZN', Timestamp('2017-10-17 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-18 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-19 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-20 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-23 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-24 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-25 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-26 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-27 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-30 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-10-31 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-11-01 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-11-02 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('2017-11-03 00:00:00')) Successfully.
Loaded Data ('AMZN', Timestamp('20

Loaded Data ('AAPL', Timestamp('2017-12-05 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-06 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-07 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-08 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-11 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-12 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-13 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-14 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-15 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-18 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-19 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-20 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-21 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('2017-12-22 00:00:00')) Successfully.
Loaded Data ('AAPL', Timestamp('20

Loaded Data ('NFLX', Timestamp('2017-11-22 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-11-24 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-11-27 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-11-28 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-11-29 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-11-30 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-01 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-04 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-05 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-06 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-07 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-08 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-11 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2017-12-12 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('20

Loaded Data ('NFLX', Timestamp('2020-08-14 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-17 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-18 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-19 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-20 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-21 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-24 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-25 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-26 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-27 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-28 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-08-31 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-09-01 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('2020-09-02 00:00:00')) Successfully.
Loaded Data ('NFLX', Timestamp('20

Loaded Data ('GOOGL', Timestamp('2020-03-02 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-03 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-04 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-05 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-06 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-09 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-10 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-11 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-12 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-13 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-16 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-17 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-18 00:00:00')) Successfully.
Loaded Data ('GOOGL', Timestamp('2020-03-19 00:00:00')) Successfully.
Loaded Data ('GOOGL'

In [22]:
# Demonstrating the 3347th stock object stored in the stock object list 
pick = 0

print("Ticker: {}".format(faang_list[pick].index))
print("Opening price: ${:.2f}".format(faang_list[pick].openp))
print("The highest price: ${:.2f}".format(faang_list[pick].highp))
print("The lowest price: ${:.2f}".format(faang_list[pick].lowp))
print("The closing price: ${:.2f}".format(faang_list[pick].closep))
print("The adjusted closing price: ${:.2f}".format(faang_list[pick].adj_close))
print("The volume: ${:.0f}".format(faang_list[pick].vol))

Ticker: ('FB', Timestamp('2017-09-29 00:00:00'))
Opening price: $168.83
The highest price: $171.66
The lowest price: $168.81
The closing price: $170.87
The adjusted closing price: $170.87
The volume: $15340400


In [23]:
faang_list[0:10]

In [24]:
faang.iloc[1]

High         1.719e+02
Low          1.688e+02
Open         1.714e+02
Close        1.695e+02
Volume       1.339e+07
Adj Close    1.695e+02
Name: (FB, 2017-10-02 00:00:00), dtype: float64

### Q5 Run the iteration procedure and for each security, save the data into a dictionary indexed by datetime

In [25]:
AAPL_dic = {}
for row in AAPL.iterrows():
    AAPL_dic[row[0]] = row[1]
AAPL_dic

{Timestamp('2017-09-29 00:00:00'): High         3.853e+01
 Low          3.800e+01
 Open         3.830e+01
 Close        3.853e+01
 Volume       1.052e+08
 Adj Close    3.665e+01
 Name: 2017-09-29 00:00:00, dtype: float64,
 Timestamp('2017-10-02 00:00:00'): High         3.861e+01
 Low          3.818e+01
 Open         3.856e+01
 Close        3.845e+01
 Volume       7.480e+07
 Adj Close    3.657e+01
 Name: 2017-10-02 00:00:00, dtype: float64,
 Timestamp('2017-10-03 00:00:00'): High         3.877e+01
 Low          3.848e+01
 Open         3.850e+01
 Close        3.862e+01
 Volume       6.492e+07
 Adj Close    3.673e+01
 Name: 2017-10-03 00:00:00, dtype: float64,
 Timestamp('2017-10-04 00:00:00'): High         3.847e+01
 Low          3.812e+01
 Open         3.841e+01
 Close        3.837e+01
 Volume       8.066e+07
 Adj Close    3.649e+01
 Name: 2017-10-04 00:00:00, dtype: float64,
 Timestamp('2017-10-05 00:00:00'): High         3.886e+01
 Low          3.851e+01
 Open         3.854e+01
 Close

In [26]:
GOOGL_dic = {}
for row in GOOGL.iterrows():
    GOOGL_dic[row[0]] = row[1]
GOOGL_dic

{Timestamp('2017-09-29 00:00:00'): High         9.758e+02
 Low          9.660e+02
 Open         9.660e+02
 Close        9.737e+02
 Volume       2.031e+06
 Adj Close    9.737e+02
 Name: 2017-09-29 00:00:00, dtype: float64,
 Timestamp('2017-10-02 00:00:00'): High         9.777e+02
 Low          9.620e+02
 Open         9.757e+02
 Close        9.675e+02
 Volume       1.539e+06
 Adj Close    9.675e+02
 Name: 2017-10-02 00:00:00, dtype: float64,
 Timestamp('2017-10-03 00:00:00'): High         9.724e+02
 Low          9.627e+02
 Open         9.676e+02
 Close        9.721e+02
 Volume       1.083e+06
 Adj Close    9.721e+02
 Name: 2017-10-03 00:00:00, dtype: float64,
 Timestamp('2017-10-04 00:00:00'): High         9.744e+02
 Low          9.656e+02
 Open         9.718e+02
 Close        9.668e+02
 Volume       1.058e+06
 Adj Close    9.668e+02
 Name: 2017-10-04 00:00:00, dtype: float64,
 Timestamp('2017-10-05 00:00:00'): High         9.865e+02
 Low          9.703e+02
 Open         9.728e+02
 Close

In [27]:
NFLX_dic = {}
for row in NFLX.iterrows():
    NFLX_dic[row[0]] = row[1]
NFLX_dic

{Timestamp('2017-09-29 00:00:00'): High         1.829e+02
 Low          1.806e+02
 Open         1.807e+02
 Close        1.814e+02
 Volume       4.228e+06
 Adj Close    1.814e+02
 Name: 2017-09-29 00:00:00, dtype: float64,
 Timestamp('2017-10-02 00:00:00'): High         1.828e+02
 Low          1.766e+02
 Open         1.821e+02
 Close        1.770e+02
 Volume       7.342e+06
 Adj Close    1.770e+02
 Name: 2017-10-02 00:00:00, dtype: float64,
 Timestamp('2017-10-03 00:00:00'): High         1.797e+02
 Low          1.776e+02
 Open         1.776e+02
 Close        1.792e+02
 Volume       3.957e+06
 Adj Close    1.792e+02
 Name: 2017-10-03 00:00:00, dtype: float64,
 Timestamp('2017-10-04 00:00:00'): High         1.867e+02
 Low          1.812e+02
 Open         1.814e+02
 Close        1.844e+02
 Volume       9.236e+06
 Adj Close    1.844e+02
 Name: 2017-10-04 00:00:00, dtype: float64,
 Timestamp('2017-10-05 00:00:00'): High         1.945e+02
 Low          1.845e+02
 Open         1.856e+02
 Close

In [28]:
FB_dic = {}
for row in FB.iterrows():
    FB_dic[row[0]] = row[1]
FB_dic

{Timestamp('2017-09-29 00:00:00'): High         1.717e+02
 Low          1.688e+02
 Open         1.688e+02
 Close        1.709e+02
 Volume       1.534e+07
 Adj Close    1.709e+02
 Name: 2017-09-29 00:00:00, dtype: float64,
 Timestamp('2017-10-02 00:00:00'): High         1.719e+02
 Low          1.688e+02
 Open         1.714e+02
 Close        1.695e+02
 Volume       1.339e+07
 Adj Close    1.695e+02
 Name: 2017-10-02 00:00:00, dtype: float64,
 Timestamp('2017-10-03 00:00:00'): High         1.707e+02
 Low          1.692e+02
 Open         1.693e+02
 Close        1.700e+02
 Volume       8.190e+06
 Adj Close    1.700e+02
 Name: 2017-10-03 00:00:00, dtype: float64,
 Timestamp('2017-10-04 00:00:00'): High         1.707e+02
 Low          1.683e+02
 Open         1.698e+02
 Close        1.684e+02
 Volume       1.213e+07
 Adj Close    1.684e+02
 Name: 2017-10-04 00:00:00, dtype: float64,
 Timestamp('2017-10-05 00:00:00'): High         1.713e+02
 Low          1.686e+02
 Open         1.692e+02
 Close

In [29]:
AMZN_dic = {}
for row in AMZN.iterrows():
    AMZN_dic[row[0]] = row[1]
AMZN_dic 

{Timestamp('2017-09-29 00:00:00'): High         9.648e+02
 Low          9.584e+02
 Open         9.601e+02
 Close        9.613e+02
 Volume       2.544e+06
 Adj Close    9.613e+02
 Name: 2017-09-29 00:00:00, dtype: float64,
 Timestamp('2017-10-02 00:00:00'): High         9.673e+02
 Low          9.521e+02
 Open         9.640e+02
 Close        9.592e+02
 Volume       2.443e+06
 Adj Close    9.592e+02
 Name: 2017-10-02 00:00:00, dtype: float64,
 Timestamp('2017-10-03 00:00:00'): High         9.637e+02
 Low          9.504e+02
 Open         9.580e+02
 Close        9.571e+02
 Volume       2.667e+06
 Adj Close    9.571e+02
 Name: 2017-10-03 00:00:00, dtype: float64,
 Timestamp('2017-10-04 00:00:00'): High         9.678e+02
 Low          9.540e+02
 Open         9.542e+02
 Close        9.655e+02
 Volume       2.527e+06
 Adj Close    9.655e+02
 Name: 2017-10-04 00:00:00, dtype: float64,
 Timestamp('2017-10-05 00:00:00'): High         9.815e+02
 Low          9.696e+02
 Open         9.700e+02
 Close

## Part 2

In [30]:
import pymongo
from pymongo import MongoClient

In [31]:
pymongo.__version__

'3.11.0'

In [32]:
#!pip install dnspython

In [33]:
#!pip3 install pymongo[srv]
#!pip3 install pymongo[tls]

### Q2.1 Connect to MongoDB

In [34]:
cluster = MongoClient("mongodb+srv://Yyq:nff4KQ2JVQdivRjU@cluster0.g8u7p.mongodb.net/<dbname>?retryWrites=true&w=majority")

### Q2.2 Access Sample Data in MongDB

In [35]:
#access the "analytics" collection and the accounts database 
db = cluster.sample_analytics
accounts=db.accounts
accounts

Collection(Database(MongoClient(host=['cluster0-shard-00-01.g8u7p.mongodb.net:27017', 'cluster0-shard-00-02.g8u7p.mongodb.net:27017', 'cluster0-shard-00-00.g8u7p.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-ow8vpt-shard-0', ssl=True), 'sample_analytics'), 'accounts')

In [36]:
#access the products for the "analytics"database
info=accounts.find_one() #import all the data
print(info.keys()) #see what contains, note it contains reviews
product_db=pd.DataFrame(info["products"]) #extract the reviews
product_db.head(7)

dict_keys(['_id', 'account_id', 'limit', 'products'])


                 0
0      Derivatives
1  InvestmentStock

In [37]:
#access the "Airbnb" collection and the listAndReviews database 
airbnb_db = cluster.sample_airbnb
review_db=airbnb_db.listingsAndReviews
#acces the reviews for the database
review=review_db.find_one() #import all the data
print(review.keys()) #see what contains, note it contains reviews


dict_keys(['_id', 'listing_url', 'name', 'summary', 'space', 'description', 'neighborhood_overview', 'notes', 'transit', 'access', 'interaction', 'house_rules', 'property_type', 'room_type', 'bed_type', 'minimum_nights', 'maximum_nights', 'cancellation_policy', 'last_scraped', 'calendar_last_scraped', 'first_review', 'last_review', 'accommodates', 'bedrooms', 'beds', 'number_of_reviews', 'bathrooms', 'amenities', 'price', 'security_deposit', 'cleaning_fee', 'extra_people', 'guests_included', 'images', 'host', 'address', 'availability', 'review_scores', 'reviews'])


In [38]:
reviews=pd.DataFrame(review["reviews"]) #extract the reviews
reviews.head(7)

        _id                date listing_id reviewer_id reviewer_name  \
0  58663741 2016-01-03 05:00:00   10006546    51483096         Cátia   
1  62413197 2016-02-14 05:00:00   10006546    40031996          Théo   
2  68310569 2016-04-04 04:00:00   10006546    53859850          Bart   
3  69693942 2016-04-12 04:00:00   10006546     3135623          Alex   
4  71451096 2016-04-25 04:00:00   10006546    13187858           Dan   
5  73191268 2016-05-07 04:00:00   10006546    61434690          Anne   
6  77438107 2016-05-31 04:00:00   10006546     3673216         Chris   

                                            comments  
0  A casa da Ana e do Gonçalo foram o local escol...  
1  We are french's students, we traveled some day...  
2  We had a spledid time in the old centre of Por...  
3  Ana and Goncalo were very helpful and accommod...  
4  The hosts were very accommodating and agreed t...  
5  THE PLACE TO BE, l'appartement est situé au mi...  
6  We had a great time in Porto. Ana’s

### Q2.3 Create a notebook to extract data

In [39]:
#Load collection first
new_db = cluster.New
new_db

Database(MongoClient(host=['cluster0-shard-00-01.g8u7p.mongodb.net:27017', 'cluster0-shard-00-02.g8u7p.mongodb.net:27017', 'cluster0-shard-00-00.g8u7p.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-ow8vpt-shard-0', ssl=True), 'New')

In [40]:
#load fruit database
fruit = new_db.fruit

#create instances for fruit database
apple = { "_id": "apples", "qty": 5 }
banana = { "_id": "bananas", "qty": 7 }
orange = { "_id": "oranges", "qty": { "in stock": 8, "ordered": 12 } }
avocado = { "_id": "avocados", "qty": "fourteen" }

#insert the dict in the database 
#fruit.insert_many([apple,orange,banana,avocado])

BulkWriteError: batch op errors occurred, full error: {'writeErrors': [{'index': 0, 'code': 11000, 'keyPattern': {'_id': 1}, 'keyValue': {'_id': 'apples'}, 'errmsg': 'E11000 duplicate key error collection: New.fruit index: _id_ dup key: { _id: "apples" }', 'op': {'_id': 'apples', 'qty': 5}}], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 0, 'nModified': 0, 'nRemoved': 0, 'upserted': []}

In [41]:
#load the data from fruit database
find_apple = fruit.find({"_id":"apples"})

In [42]:
for a in find_apple:
    print(a)

{'_id': 'apples', 'qty': 5}
